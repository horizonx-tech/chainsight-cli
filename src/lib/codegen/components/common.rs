use std::{fs::OpenOptions, path::Path, io::Read};

use proc_macro2::TokenStream;
use serde::{Deserialize, Serialize};

use crate::types::ComponentType;

#[derive(Deserialize, Serialize, Clone, Copy, Debug, PartialEq, clap::ValueEnum)]
pub enum DatasourceType {
    #[serde(rename = "canister")]
    Canister,
    #[serde(rename = "contract")]
    Contract,
}
#[derive(Deserialize, Serialize, Clone, Copy, Debug, PartialEq, clap::ValueEnum)]
pub enum DestinactionType {
    #[serde(rename = "uint256")]
    Uint256Oracle,
    #[serde(rename = "string")]
    StringOracle,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct Datasource {
    #[serde(rename = "type")]
    pub type_: DatasourceType,
    pub id: String,
    pub method: DatasourceMethod
}
#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct DatasourceMethod {
    pub identifier: String,
    pub interface: Option<String>,
    pub args: Vec<serde_yaml::Value>,
}
#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct DatasourceMethodCustomStruct {
    pub name: String,
    pub fields: Vec<DatasourceMethodCustomStructField>
}
#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct DatasourceMethodCustomStructField {
    pub name: String,
    #[serde(rename = "type")]
    pub type_: String
}
#[derive(Clone, Debug, Deserialize, Serialize)]

pub struct DatasourceMethodCustomType {
    pub name: String,
    pub types: Vec<String>
}

impl Datasource {
    // temp: use Default trait
    pub fn default_contract() -> Self {
        Self {
            type_: DatasourceType::Contract,
            id: "0000000000000000000000000000000000000000".to_string(), // temp
            method: DatasourceMethod {
                identifier: "totalSupply():(uint256)".to_string(),
                interface: Some("ERC20.json".to_string()),
                args: vec![],
            },
        }
    }
    pub fn new_contract(
        identifier: String,
        interface: Option<String>,
    ) -> Self {
        Self {
            type_: DatasourceType::Contract,
            id: "0000000000000000000000000000000000000000".to_string(), // temp
            method: DatasourceMethod {
                identifier,
                interface,
                args: vec![],
            },
        }
    }
    
    // temp: use Default trait
    pub fn default_canister(ident_with_ts: bool) -> Self {
        let identifier = if ident_with_ts {
            "get_last_snapshot : () -> (record { value : text; timestamp : nat64 })"
        } else {
            "get_last_snapshot_value : () -> (text)"
        }.to_string();

        Self {
            type_: DatasourceType::Canister,
            id: "xxxxx-xxxxx-xxxxx-xxxxx-xxx".to_string(), // temp
            method: DatasourceMethod {
                identifier,
                interface: None,
                args: vec![],
            },
        }
    }

    pub fn new_canister(
        identifier: String,
        interface: Option<String>,
    ) -> Self {
        Self {
            type_: DatasourceType::Canister,
            id: "xxxxx-xxxxx-xxxxx-xxxxx-xxx".to_string(), // temp
            method: DatasourceMethod {
                identifier,
                interface,
                args: vec![],
            },
        }
    }
}

/// Determine indexer type from manifest
pub trait ComponentManifest: std::fmt::Debug {
    fn load(path: &str) -> anyhow::Result<Self> where Self: Sized;
    fn to_str_as_yaml(&self) -> anyhow::Result<String> where Self: Sized;
    fn validate_manifest(&self) -> anyhow::Result<()>;
    fn generate_codes(&self, interface_contract: Option<ethabi::Contract>) -> anyhow::Result<TokenStream>;

    fn component_type(&self) -> ComponentType;
    fn label(&self) -> &str;
    fn destination_type(&self) -> Option<DestinactionType>;
    fn required_interface(&self) -> Option<String>;
}

#[derive(Deserialize)]
pub struct ComponentTypeInManifest {
    #[serde(rename = "type")]
    pub type_: ComponentType,
}
impl ComponentTypeInManifest {
    pub fn determine_type(component_manifest_path: &str) -> anyhow::Result<ComponentType> {
        let mut file = OpenOptions::new()
            .read(true)
            .open(&Path::new(component_manifest_path))?;
        let mut contents = String::new();
        file.read_to_string(&mut contents)?;
        let data: Self = serde_yaml::from_str(&contents)?;
        Ok(data.type_)
    }
}
