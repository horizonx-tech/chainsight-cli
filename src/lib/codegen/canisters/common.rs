use quote::{format_ident, quote};

use crate::lib::codegen::components::common::{DatasourceMethodCustomStruct, DatasourceMethodCustomType, DatasourceMethodArg};

pub enum OutsideCallIdentsType {
    Eth,
    CrossCanisterCall,
    All
}
/// Generate common identifies such as storage, setter, etc. for outside calls
pub fn generate_outside_call_idents(type_: OutsideCallIdentsType) -> proc_macro2::TokenStream {
    let eth_idents = quote! {
        define_web3_ctx!();
        define_transform_for_web3!();

        manage_single_state!("target_addr", String, false);
    };
    let cross_canister_call_idents = quote! {
        manage_single_state!("target_canister", String, false);
    };
    match type_ {
        OutsideCallIdentsType::Eth => {
            quote! {
                #eth_idents

                setup_func!({
                    target_addr: String,
                    web3_ctx_param: Web3CtxParam
                });
            }
        },
        OutsideCallIdentsType::CrossCanisterCall => {
            quote! {
                #cross_canister_call_idents

                setup_func!({
                    target_canister: String
                });
            }
        },
        OutsideCallIdentsType::All => {
            quote! {
                #eth_idents
                #cross_canister_call_idents

                setup_func!({
                    target_canister: String,
                    target_addr: String,
                    web3_ctx_param: Web3CtxParam
                });
            }
        }
    }
}

// Generate the part of data of the argument that calls the function of datasource contract/canister
pub fn generate_request_arg_idents(method_args: &Vec<DatasourceMethodArg>) -> (Vec<proc_macro2::TokenStream>, Vec<proc_macro2::Ident>) {
    let mut value_idents = vec![];
    let mut type_idents = vec![];
    for method_arg in method_args {
        let DatasourceMethodArg { type_, value } = method_arg;
        // temp
        let request_arg_value = match type_.clone().as_str() {
            "ic_web3::types::U256" => {
                match value {
                    serde_yaml::Value::String(val) => quote! { ic_web3::types::U256::from_dec_str(#val).unwrap() },
                    serde_yaml::Value::Number(val) => {
                        match val.as_u64() {
                            Some(val) => quote! { #val.into() },
                            None => quote! {}
                        }
                    },
                    _ => quote! {}
                }
            }
            "ic_web3::types::Address" => {
                match value {
                    serde_yaml::Value::String(val) => quote! { ic_web3::types::Address::from_str(#val).unwrap() },
                    _ => quote! {}
                }
            },
            _ => {
                match value {
                    serde_yaml::Value::String(val) => {
                        quote! { #val, }
                    },
                    serde_yaml::Value::Number(val) => {
                        match val.as_u64() {
                            Some(val) => {
                                let type_ident = format_ident!("{}", type_);
                                quote! { #val as #type_ident }
                            },
                            None => {
                                quote! {}
                            }
                        }
                    },
                    _ => {
                        quote! {}
                    }
                }
            }
        };
        value_idents.push(request_arg_value);
        if type_ == "ic_web3::types::U256" || type_ == "ic_web3::types::Address" {
            // In the case of contract, other than the primitive type (ic_web3::types::U256 etc.) may be set, in which case type_idents is not used.
            type_idents.push(format_ident!("String")); // temp: thread 'main' panicked at '"ic_web3::types::U256" is not a valid Ident'
        } else {
            type_idents.push(format_ident!("{}", type_));
        }
    };
    (value_idents, type_idents)
}

// Generate CustomStruct Identifiers from manifest's struct data
pub fn generate_custom_struct_idents(custom_structs: &Vec<DatasourceMethodCustomStruct>) -> Vec<proc_macro2::TokenStream> {
    let mut custom_struct_ident: Vec<proc_macro2::TokenStream> = vec![];
    for custom_struct_def in custom_structs {
        let struct_ident = format_ident!("{}", &custom_struct_def.name);
        let mut custom_struct_fields = vec![];
        for field in &custom_struct_def.fields {
            let field_name_ident = format_ident!("{}", &field.name);
            let field_type_ident = format_ident!("{}", &field.type_);
            custom_struct_fields.push(quote! {
                pub #field_name_ident: #field_type_ident,
            });
        }
        custom_struct_ident.push(quote! {
            #[derive(Debug, Clone, candid::CandidType, candid::Deserialize)]
            pub struct #struct_ident {
                #(#custom_struct_fields)*
            }
        });
    };
    custom_struct_ident
}

// Generate CustomType Identifiers from manifest's struct data
pub fn generate_custom_type_idents(custom_types: &Vec<DatasourceMethodCustomType>) -> Vec<proc_macro2::TokenStream> {
    let mut custom_type_ident: Vec<proc_macro2::TokenStream> = vec![];
    for custom_type_def in custom_types {
        let type_ident = format_ident!("{}", &custom_type_def.name);
        let mut custom_type_scalars = vec![];
        for type_ in &custom_type_def.types {
            custom_type_scalars.push(format_ident!("{}", &type_));
        }
        custom_type_ident.push(quote! {
            type #type_ident = (#(#custom_type_scalars),*);
        });
    }
    custom_type_ident
}