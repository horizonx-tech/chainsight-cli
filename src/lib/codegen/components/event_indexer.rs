use std::collections::HashMap;

use anyhow::bail;
use serde::{Deserialize, Serialize};
use serde_json::json;

use crate::{
    lib::codegen::{canisters, scripts},
    types::{ComponentType, Network},
};

use super::common::{
    custom_tags_interval_sec, ComponentManifest, ComponentMetadata, GeneratedCodes, SourceType,
    Sources, DEFAULT_MONITOR_DURATION_SECS,
};

/// Component Manifest: Event Indexer
#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct EventIndexerComponentManifest {
    #[serde(skip_serializing)]
    pub id: Option<String>,
    pub version: String,
    pub metadata: ComponentMetadata,
    pub datasource: EventIndexerDatasource,
    pub interval: u32,
}

impl From<EventIndexerComponentManifest>
    for chainsight_cdk::config::components::EventIndexerConfig
{
    fn from(val: EventIndexerComponentManifest) -> Self {
        Self {
            common: chainsight_cdk::config::components::CommonConfig {
                canister_name: val.id.clone().unwrap(),
                monitor_duration: DEFAULT_MONITOR_DURATION_SECS,
            },
            def: chainsight_cdk::config::components::EventIndexerEventDefinition {
                identifier: val.datasource.event.identifier,
                abi_file_path: format!(
                    "./__interfaces/{}",
                    val.datasource.event.interface.unwrap()
                ),
            },
        }
    }
}

impl EventIndexerComponentManifest {
    pub fn new(
        id: &str,
        label: &str,
        description: &str,
        version: &str,
        datasource: EventIndexerDatasource,
        interval: u32,
    ) -> Self {
        Self {
            id: Some(id.to_owned()),
            version: version.to_owned(),
            metadata: ComponentMetadata {
                label: label.to_owned(),
                type_: ComponentType::EventIndexer,
                description: description.to_owned(),
                tags: Some(vec![
                    "Ethereum".to_string(),
                    "ERC-20".to_string(),
                    "Transfer".to_string(),
                ]),
            },
            datasource,
            interval,
        }
    }
}
impl ComponentManifest for EventIndexerComponentManifest {
    fn load_with_id(path: &str, id: &str) -> anyhow::Result<Self> {
        let manifest = Self::load(path)?;
        Ok(Self {
            id: Some(id.to_owned()),
            ..manifest
        })
    }

    fn to_str_as_yaml(&self) -> anyhow::Result<String> {
        let yaml = serde_yaml::to_string(&self)?;
        Ok(self.yaml_str_with_configs(yaml, "event_indexer".to_string()))
    }

    fn validate_manifest(&self) -> anyhow::Result<()> {
        canisters::event_indexer::validate_manifest(self)
    }

    fn generate_codes(
        &self,
        _interface_contract: Option<ethabi::Contract>,
    ) -> anyhow::Result<GeneratedCodes> {
        Ok(GeneratedCodes {
            lib: canisters::event_indexer::generate_codes(self)?,
            types: None,
        })
    }

    fn generate_scripts(&self, network: Network) -> anyhow::Result<String> {
        scripts::event_indexer::generate_scripts(self, network)
    }

    fn component_type(&self) -> ComponentType {
        ComponentType::EventIndexer
    }

    fn id(&self) -> Option<String> {
        self.id.clone()
    }

    fn metadata(&self) -> &ComponentMetadata {
        &self.metadata
    }

    fn destination_type(&self) -> Option<super::common::DestinationType> {
        None
    }
    fn get_sources(&self) -> Sources {
        #[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
        struct Attributes {
            chain_id: u64,
            event_name: String,
            contract_type: String,
        }
        let contract_type = self.clone().datasource.contract_type.unwrap_or_else(|| {
            self.datasource
                .event
                .interface
                .clone()
                .map(|s| s.replace(".json", ""))
                .unwrap()
        });

        let mut attr = HashMap::new();
        attr.insert(
            "chain_id".to_string(),
            json!(self.datasource.network.chain_id),
        );
        attr.insert(
            "event_name".to_string(),
            json!(self.datasource.event.identifier),
        );
        attr.insert("contract_type".to_string(), json!(contract_type));

        Sources {
            source_type: SourceType::Evm,
            source: self.datasource.clone().id,
            attributes: attr,
        }
    }

    fn required_interface(&self) -> Option<String> {
        self.datasource.event.interface.clone()
    }

    fn generate_user_impl_template(&self) -> anyhow::Result<GeneratedCodes> {
        bail!("not implemented")
    }
    fn custom_tags(&self) -> HashMap<String, String> {
        let mut res = HashMap::new();
        let (interval_key, interval_val) = custom_tags_interval_sec(self.interval);
        res.insert(interval_key, interval_val);
        res
    }
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct EventIndexerDatasource {
    pub id: String,
    pub event: EventIndexerEventDefinition,
    pub network: SourceNetwork,
    pub from: u64,
    pub contract_type: Option<String>,
}
#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct SourceNetwork {
    pub rpc_url: String,
    pub chain_id: u64,
}

impl EventIndexerDatasource {
    pub fn new(
        id: String,
        event: EventIndexerEventDefinition,
        network: SourceNetwork,
        from: u64,
        contract_type: Option<String>,
    ) -> Self {
        Self {
            id,
            event,
            network,
            from,
            contract_type,
        }
    }

    pub fn default() -> Self {
        Self {
            id: "0x6B175474E89094C44Da98b954EedeAC495271d0F".to_string(),
            event: EventIndexerEventDefinition::new(
                "Transfer".to_string(),
                Some("ERC20.json".to_string()),
            ),
            network: SourceNetwork {
                rpc_url: "https://mainnet.infura.io/v3/${INFURA_MAINNET_RPC_URL_KEY}".to_string(),
                chain_id: 1,
            },
            from: 17660942,
            contract_type: Some("ERC-20".to_string()),
        }
    }
}
#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct EventIndexerEventDefinition {
    pub identifier: String,
    pub interface: Option<String>,
}
impl EventIndexerEventDefinition {
    pub fn new(identifier: String, interface: Option<String>) -> Self {
        Self {
            identifier,
            interface,
        }
    }
}

#[cfg(test)]
mod tests {
    use std::fs::File;

    use insta::assert_display_snapshot;
    use jsonschema::JSONSchema;

    use crate::lib::test_utils::SrcString;

    use super::*;

    #[test]
    fn test_to_manifest_struct() {
        let yaml = r#"
version: v1
metadata:
    label: sample_event_indexer
    type: event_indexer
    description: Description
    tags:
    - Ethereum
    - ERC-20
    - Transfer
datasource:
    id: 0x6B175474E89094C44Da98b954EedeAC495271d0F
    event:
        identifier: Transfer
        interface: ERC20.json
    contract_type: ERC20
    network: 
        rpc_url: https://mainnet.infura.io/v3/${INFURA_MAINNET_RPC_URL_KEY}
        chain_id: 1
    from: 17660942
interval: 3600
        "#;

        let result = serde_yaml::from_str::<EventIndexerComponentManifest>(yaml);
        assert!(result.is_ok());
        let component = result.unwrap();
        assert_eq!(
            component,
            EventIndexerComponentManifest {
                id: None,
                version: "v1".to_string(),
                metadata: ComponentMetadata {
                    label: "sample_event_indexer".to_string(),
                    type_: ComponentType::EventIndexer,
                    description: "Description".to_string(),
                    tags: Some(vec![
                        "Ethereum".to_string(),
                        "ERC-20".to_string(),
                        "Transfer".to_string()
                    ])
                },
                datasource: EventIndexerDatasource {
                    id: "0x6B175474E89094C44Da98b954EedeAC495271d0F".to_string(),
                    event: EventIndexerEventDefinition {
                        identifier: "Transfer".to_string(),
                        interface: Some("ERC20.json".to_string())
                    },
                    network: SourceNetwork {
                        rpc_url: "https://mainnet.infura.io/v3/${INFURA_MAINNET_RPC_URL_KEY}"
                            .to_string(),
                        chain_id: 1,
                    },
                    from: 17660942,
                    contract_type: Some("ERC20".to_string())
                },
                interval: 3600
            }
        );
        let schema = serde_json::from_str(include_str!(
            "../../../../resources/schema/event_indexer.json"
        ))
        .expect("Invalid json");
        let instance = serde_yaml::from_str(yaml).expect("Invalid yaml");
        let compiled = JSONSchema::compile(&schema).expect("Invalid schema");
        let result = compiled.validate(&instance);
        assert!(result.is_ok());
    }

    #[test]
    fn test_snapshot_outputs() {
        let manifest = EventIndexerComponentManifest {
            id: Some("sample_event_indexer".to_string()),
            version: "v1".to_string(),
            metadata: ComponentMetadata {
                label: "Sample Event Indexer".to_string(),
                type_: ComponentType::EventIndexer,
                description: "Description".to_string(),
                tags: Some(vec![
                    "Ethereum".to_string(),
                    "ERC-20".to_string(),
                    "Transfer".to_string(),
                ]),
            },
            datasource: EventIndexerDatasource {
                id: "0x6B175474E89094C44Da98b954EedeAC495271d0F".to_string(),
                event: EventIndexerEventDefinition {
                    identifier: "Transfer".to_string(),
                    interface: Some("ERC20.json".to_string()),
                },
                network: SourceNetwork {
                    rpc_url: "https://mainnet.infura.io/v3/${INFURA_MAINNET_RPC_URL_KEY}"
                        .to_string(),
                    chain_id: 1,
                },
                from: 17660942,
                contract_type: Some("ERC20".to_string()),
            },
            interval: 3600,
        };

        let abi = File::open("resources/ERC20.json").unwrap();
        let generated_codes = manifest
            .generate_codes(Option::Some(ethabi::Contract::load(abi).unwrap()))
            .unwrap();
        assert_display_snapshot!(SrcString::from(&generated_codes.lib));
        assert!(generated_codes.types.is_none());

        assert!(manifest.generate_user_impl_template().is_err());

        assert_display_snapshot!(&manifest.generate_scripts(Network::Local).unwrap());
    }
}
