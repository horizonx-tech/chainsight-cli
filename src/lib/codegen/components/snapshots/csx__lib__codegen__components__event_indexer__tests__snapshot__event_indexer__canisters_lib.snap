---
source: src/lib/codegen/components/event_indexer.rs
expression: "SrcString::from(generated_codes.lib)"
---
use chainsight_cdk_macros::def_event_indexer_canister;
def_event_indexer_canister!("{
\"common\":{
\"canister_name\":\"sample_event_indexer\"
}, \"def\":{
\"identifier\":\"Transfer\", \"abi_file_path\":\"./__interfaces/ERC20.json\"
}}");

